### PR TITLE
Improve SignalR Nginx settings

### PR DESCRIPTION
**Pull request by BrennanConroy**
Progress on https://github.com/dotnet/AspNetCore.Docs/issues/10827

### CODE REVIEW

File: aspnetcore/signalr/scale.md
Comment (halter73) [inline aspnetcore/signalr/scale.md:114]: What's going on here?

Comment (BrennanConroy) [reply aspnetcore/signalr/scale.md:114]: Like I was saying on the issue, "Connection: close" is applied by default in Nginx, and "Connection: keep-alive" should not be forwarded by proxies as it is a hop-by-hop header. We can have another discussion about that behavior, but I think for this PR this is a big improvement over what we had before.

File: aspnetcore/signalr/scale.md
Comment (halter73) [inline aspnetcore/signalr/scale.md:135]: It looks like this is needed for WebSockets if the KeepAliveInterval is large enough http://nginx.org/en/docs/http/websocket.html

File: aspnetcore/signalr/scale.md
Comment (halter73) [reply aspnetcore/signalr/scale.md:114]: Now I remember. I have the same question [I had on the issue](https://github.com/dotnet/AspNetCore.Docs/issues/10827#issuecomment-631177400).

> Does this mean that every request to /hubroute will always contain either a "Connection: upgrade" or a "Connection: close" header? We wouldn't want to disable keep-alive for long polling.

Sure, "Connection" is a hop-by-hop header, but I want to make sure the connection between nginx and the signalr server isn't closed after every request.

File: aspnetcore/signalr/scale.md
Comment (halter73) [inline aspnetcore/signalr/scale.md:130]: If we worried about the proxy caching requests, shouldn't we be worried about every request, not just WebSocket requests?

```suggestion
      proxy_cache off;
```

File: aspnetcore/signalr/scale.md
Comment (halter73) [inline aspnetcore/signalr/scale.md:120]: Why bother with this in the signalr doc? What if wwwroot is on another machine? I expect static files will be proxied in many cases.

File: aspnetcore/signalr/scale.md
Comment (guardrex) [inline aspnetcore/signalr/scale.md:146]: I think you meant to combine sentences into a paragraph ...

```suggestion
When multiple backend servers are used, sticky sessions must be added to prevent SignalR connections from switching servers when connecting. There are multiple ways to add sticky sessions in Nginx. Two approaches are shown below depending on what you have available.

```

File: aspnetcore/signalr/scale.md
Comment (guardrex) [inline aspnetcore/signalr/scale.md:147]:
```suggestion
```

File: aspnetcore/signalr/scale.md
Comment (guardrex) [inline aspnetcore/signalr/scale.md:168]:
```suggestion
Finally, change `proxy_pass http://localhost:5000` in the `server` section to `proxy_pass http://backend`.
```

File: aspnetcore/signalr/scale.md
Comment (guardrex) [inline aspnetcore/signalr/scale.md:170]: Turns out for this link, they don't redirect, so we can go slash-less on it ...

```suggestion
For more information on WebSockets over Nginx, see [NGINX as a WebSocket Proxy](https://www.nginx.com/blog/websocket-nginx).
```

File: aspnetcore/signalr/scale.md
Comment (guardrex) [inline aspnetcore/signalr/scale.md:148]: Recommend moving as much of the unlocalized language into the text as possible.

```suggestion
The following is added in addition to the previous configuration. `backend` in the following example is the name of the group of servers:
```

File: aspnetcore/signalr/scale.md
Comment (guardrex) [inline aspnetcore/signalr/scale.md:152]:
```suggestion
```

**Review by guardrex — CHANGES_REQUESTED** [review]
WRT "Nginx Plus" and "Nginx Open Source" ... If those aren't really meant to be used together and given our general desire to place text in the doc where it will MT, I recommend splitting into two examples ... one for each. The lead-in line(s) can then be in text and translate. If you want that but don't want to deal with the setup, I can edit directly for you to sign-off on. :ear:

File: aspnetcore/signalr/scale.md
Comment (guardrex) [inline aspnetcore/signalr/scale.md:146]:
```suggestion
The following is added in addition to the previous configuration. In the following examples, `backend` is the name of the group of servers.
```

*"Ooops!*" on this one. The conditional is better up-front.

File: aspnetcore/signalr/scale.md
Comment (guardrex) [inline aspnetcore/signalr/scale.md:148]:
```suggestion
With [Nginx Open Source](https://nginx.org/en/), use `ip_hash` to route connections to a server based on the client's IP address:
```

File: aspnetcore/signalr/scale.md
Comment (guardrex) [inline aspnetcore/signalr/scale.md:163]:
```suggestion
With [Nginx Plus](https://www.nginx.com/products/nginx), use `sticky` to add a cookie to requests and pin the user's requests to a server:
```

**Review by guardrex — APPROVED** [review]
Just some nit suggestions.

File: aspnetcore/signalr/scale.md
Comment (guardrex) [reply aspnetcore/signalr/scale.md:148]: btw - the "en" localization is required for their doc to load. **_Shame on them!_** :smile: lol

**Comment by guardrex** [comment]
@BrennanConroy ... I made a few final nit suggestions :point_up:. Outside of those, was there anything else you wanted to do on this one before merging it?

**Comment by BrennanConroy** [comment]
> Outside of those, was there anything else you wanted to do on this one before merging it?

Yeah, need to figure out what to do about keep-alive

**Comment by guardrex** [comment]
Ok ... I'll leave it to you to merge when ready.

File: aspnetcore/signalr/scale.md
Comment (BrennanConroy) [reply aspnetcore/signalr/scale.md:114]: Is it safe to make all connections keep alive?

Comment (halter73) [reply aspnetcore/signalr/scale.md:114]: I don't see why not. Most modern HTTP proxies and clients do. It can be problem with stateful NTLM connections, but we're not telling people how to configure NTLM.

@Tratcher

Comment (Tratcher) [reply aspnetcore/signalr/scale.md:114]: Yes, enable keep-alive by default between nginx and kestrel.

File: aspnetcore/signalr/scale.md
Comment (BrennanConroy) [inline aspnetcore/signalr/scale.md:113]:
```suggestion
    '' keep-alive;
```

**Comment by Tratcher** [comment]
How different is this guidance from our general nginx guidance? Should we be improving that instead and linking to it from here with signalr specific notes?

**Comment by BrennanConroy** [comment]
> How different is this guidance from our general nginx guidance? Should we be improving that instead and linking to it from here with signalr specific notes?

I prefer cross-referencing the docs.

File: aspnetcore/signalr/scale.md
Comment (halter73) [inline aspnetcore/signalr/scale.md:113]: What's the reason for basing the Connection header on the incoming Upgrade header? Maybe the following which bases it on the incoming Connection header would make more sense. It seems more straightforward not to look at a different header.

```
map $http_connection $connection_upgrade {
  Upgrade Upgrade;
  default keep-alive;
}
```

Comment (BrennanConroy) [reply aspnetcore/signalr/scale.md:113]: Doesn't work if the request has a multi-value Connection header.

Comment (BrennanConroy) [reply aspnetcore/signalr/scale.md:113]: Looks like it might work if I do `"~*Upgrade" Upgrade;`

Comment (halter73) [reply aspnetcore/signalr/scale.md]: Given that, I'm fine with the way you have it now or the old way.

**Review by halter73 — APPROVED** [review]

File: aspnetcore/signalr/scale.md
Comment (halter73) [inline aspnetcore/signalr/scale.md:112]: Out of curiosity, does the following work?

```suggestion
  "~*Upgrade" $http_connection;
```

**Comment by halter73** [comment]
We should wait for @Tratcher's approval before merging though.

File: aspnetcore/signalr/scale.md
Comment (Tratcher) [reply aspnetcore/signalr/scale.md]: What does `"~*Upgrade" Upgrade;` mean?

File: aspnetcore/signalr/scale.md
Comment (Tratcher) [inline aspnetcore/signalr/scale.md:113]: Comparing to the other nginx doc, why do `default keep-alive;` here and `proxy_http_version 1.1; proxy_set_header   Connection keep-alive;` there?

**Comment by Tratcher** [comment]
Don't block on me, but I would like to see consistent guidance across the docs for common settings like keep-alive.

File: aspnetcore/signalr/scale.md
Comment (BrennanConroy) [reply aspnetcore/signalr/scale.md:113]: Not sure what you mean, this is how you set the Connection header to Upgrade for websockets, and keep-alive for other connections. The other doc doesn't handle websockets.

File: aspnetcore/signalr/scale.md
Comment (BrennanConroy) [reply aspnetcore/signalr/scale.md]: Case-insensitive check for "Upgrade" being in the request header. For example, Firefox sets "keep-alive, Upgrade" which wouldn't match normally.

File: aspnetcore/signalr/scale.md
Comment (Tratcher) [reply aspnetcore/signalr/scale.md:113]: Can you not have a generic nginx config that works for normal requests and websockets?

Comment (Tratcher) [reply aspnetcore/signalr/scale.md:113]: If our general nginx guidance doesn't work for websockets we at least need to call that out more clearly there. There's a weird callout there for blazor server apps, but not for general websockets.

File: aspnetcore/signalr/scale.md
Comment (BrennanConroy) [reply aspnetcore/signalr/scale.md:112]: It does

File: aspnetcore/signalr/scale.md
Comment (BrennanConroy) [inline aspnetcore/signalr/scale.md:113]:
```suggestion
    default keep-alive;
```

**Pull request merged**
